### PR TITLE
Tweak `insertKeyExists` and `deleteKeyExists`

### PR DESCRIPTION
**Pull request by sjakobi**
(This currently includes #456)

### CODE REVIEW

File: Data/HashMap/Internal.hs [hunk 1]
```diff
@@ -1162,13 +1160,13 @@ delete' h0 k0 m0 = go h0 k0 0 m0
 -- We can skip:
 --  - the key equality check on the leaf, if we reach a leaf it must be the key
 deleteKeyExists :: Int -> Hash -> k -> HashMap k v -> HashMap k v
-deleteKeyExists !collPos0 !h0 !k0 !m0 = go collPos0 h0 k0 0 m0
+deleteKeyExists !collPos0 !h0 !k0 !m0 = go collPos0 h0 k0 m0
   where
-    go :: Int -> Hash -> k -> Int -> HashMap k v -> HashMap k v
-    go !_collPos !_h !_k !_s (Leaf _ _) = Empty
-    go collPos h k s (BitmapIndexed b ary) =
+    go :: Int -> Hash -> k -> HashMap k v -> HashMap k v
```
Comment (sjakobi) [inline Data/HashMap/Internal.hs:1165]:
```suggestion
    go :: Int -> Word -> k -> HashMap k v -> HashMap k v
```

This should probably be clarified further.

File: Data/HashMap/Internal.hs [hunk 2]
```diff
@@ -1181,25 +1179,27 @@ deleteKeyExists !collPos0 !h0 !k0 !m0 = go collPos0 h0 k0 0 m0
...
+    go collPos h _ (Collision _hy v)
       | A.length v == 2
       = if collPos == 0
         then Leaf h (A.index v 1)
         else Leaf h (A.index v 0)
       | otherwise = Collision h (A.delete v collPos)
-    go !_ !_ !_ !_ Empty = Empty -- error "Internal error: deleteKeyExists empty"
+    go !_ !_ !_ Empty = Empty -- error "Internal error: deleteKeyExists empty"
+
+    shiftHash h = h `unsafeShiftR` bitsPerSubkey
```
Comment (sjakobi) [inline Data/HashMap/Internal.hs:1202]: Maybe move this to the top-level?!

File: Data/HashMap/Internal.hs [hunk 3]
```diff
@@ -1181,25 +1179,27 @@ deleteKeyExists !collPos0 !h0 !k0 !m0 = go collPos0 h0 k0 0 m0
...
+            !st' = go collPos (shiftHash h) k st
         in case st' of
             Empty ->
                 let ary' = A.delete ary i
                     bm   = fullBitmap .&. complement (1 `unsafeShiftL` i)
                 in BitmapIndexed bm ary'
             _ -> Full (A.update ary i st')
-      where i = index h s
-    go collPos h _ _ (Collision _hy v)
+      where i = index h 0
```
Comment (sjakobi) [inline Data/HashMap/Internal.hs:1193]: The zero argument is a bit ugly. Maybe introduce something like `rootIndex`?!

File: Data/HashMap/Internal.hs [hunk 4]
```diff
@@ -1181,25 +1179,27 @@ deleteKeyExists !collPos0 !h0 !k0 !m0 = go collPos0 h0 k0 0 m0
                       bIndexed = BitmapIndexed (b .&. complement m) (A.delete ary i)
                 l | isLeafOrCollision l && A.length ary == 1 -> l
                 _ -> BitmapIndexed b (A.update ary i st')
-      where m = mask h s
+      where m = mask h 0
```
Comment (sjakobi) [inline Data/HashMap/Internal.hs:1182]: Same issue here with the ugly 0 argument.

**Comment by sjakobi** [comment]
These new invariant tests are great! :)

```
      alterF
        valid:                             FAIL
          *** Failed! Falsified (after 28 tests and 26 shrinks):
          {_->[Nothing]}
          K {hash = 9, _x = A}
          fromList [(K {hash = 0, _x = A},1),(K {hash = 9, _x = B},1),(K {hash = 9, _x = A},1)]
          [Invalid (INV6_misplaced_hash 0) (SubHashPath {partialHash = 9, lengthInBits = 5})] /= [Valid]
          Use --quickcheck-replay=980663 to reproduce.
          Use -p '/Data.HashMap.Strict.alterF.valid/' to rerun this test only.
```

File: Data/HashMap/Internal.hs (hunk 1 above)
Comment (sjakobi) [reply Data/HashMap/Internal.hs:1165]: The variable name `h` should also be replaced by something less missleading.

**Comment by sjakobi** [comment]
Note for myself:

I somehow couldn't reproduce the bugs above with `cabal test -O0`. I should look into this!

**Comment by treeowl** [comment]
Oh, I'm starting to understand now. You're shifting as you go to avoid an extra argument?

**Comment by sjakobi** [comment]
It seems that right now, there are no benchmarks using this code. All benchmarks for `alterF` seem to rely on rules that rewrite the code to either `insert` or `delete`.

This may change with #404 though.

**Comment by sjakobi** [comment]
I can't find a significant speedup in the `alter` benchmarks, but not a slowdown either. The generated Core seems slightly better.

**Pull request merged**